File: src/main/java/nz/co/searchwellington/feeds/FeedItemLocalCopyDecorator.java
package nz.co.searchwellington.feeds;

import java.util.List;

import nz.co.searchwellington.model.Resource;
import nz.co.searchwellington.model.frontend.FeedNewsitemAcceptanceState;
import nz.co.searchwellington.model.frontend.FeedNewsitemForAcceptance;
import nz.co.searchwellington.model.frontend.FrontendFeedNewsitem;
import nz.co.searchwellington.repositories.HibernateResourceDAO;
import nz.co.searchwellington.repositories.SupressionDAO;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import com.google.common.collect.Lists;

@Component
public class FeedItemLocalCopyDecorator {
	
	private HibernateResourceDAO resourceDAO;
	private SupressionDAO suppressionDAO;
	
	@Autowired
	public FeedItemLocalCopyDecorator(HibernateResourceDAO resourceDAO, SupressionDAO suppressionDAO) {
		this.resourceDAO = resourceDAO;
		this.suppressionDAO = suppressionDAO;
	}
	
	public List<FeedNewsitemForAcceptance> addSupressionAndLocalCopyInformation(List<FrontendFeedNewsitem> feedNewsitems) {
		final List<FeedNewsitemForAcceptance> decoratedFeednewsitems = Lists.newArrayList();
		for (FrontendFeedNewsitem feedNewsitem : feedNewsitems) {
            decoratedFeednewsitems.add(new FeedNewsitemForAcceptance(feedNewsitem, determineCurrentAcceptanceStateOf(feedNewsitem)));
		}
		return decoratedFeednewsitems;
	}

    private FeedNewsitemAcceptanceState determineCurrentAcceptanceStateOf(FrontendFeedNewsitem feedNewsitem) {
        Integer localCopyId = null;
        boolean isSuppressed = false;

        if (feedNewsitem.getUrl() != null) {
            Resource localCopy = resourceDAO.loadResourceByUrl(feedNewsitem.getUrl());	// TODO expensive?
            if (localCopy != null) {
                localCopyId = localCopy.getId();
            }
            isSuppressed = suppressionDAO.isSupressed(feedNewsitem.getUrl());
        }
        return new FeedNewsitemAcceptanceState(localCopyId, isSuppressed);
    }

}

File: src/main/scala/nz/co/searchwellington/feeds/FeedItemLocalCopyDecorator.scala
package nz.co.searchwellington.feeds

import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.controllers.admin.AdminUrlBuilder
import nz.co.searchwellington.model.User
import nz.co.searchwellington.model.frontend.{Action, FrontendNewsitem, FrontendResource}
import nz.co.searchwellington.repositories.SuppressionDAO
import nz.co.searchwellington.repositories.mongo.MongoRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future

@Component class FeedItemLocalCopyDecorator @Autowired()(mongoRepository: MongoRepository, suppressionDAO: SuppressionDAO,
                                                         adminUrlBuilder: AdminUrlBuilder) extends ReasonableWaits {

  def withFeedItemSpecificActions(feedNewsitems: Seq[FrontendResource], loggedInUser: Option[User]): Future[Seq[FrontendResource]] = {
    def addFeedItemsActions(feedNewsitem: FrontendResource): Future[FrontendResource] = {
      val eventuallyLocalCopy = mongoRepository.getResourceByUrl(feedNewsitem.url)
      val eventuallyIsSuppressed = suppressionDAO.isSupressed(feedNewsitem.url)

      for {
        localCopy <- eventuallyLocalCopy
        isSupressed <- eventuallyIsSuppressed
      } yield {
        feedNewsitem match {
          case n: FrontendNewsitem =>
            loggedInUser.map { l =>
              val acceptOrEditAction = localCopy.map { lc =>
                Action("Edit local copy", adminUrlBuilder.getResourceEditUrl(lc))
              }.getOrElse(
                Action("Accept", adminUrlBuilder.getFeednewsItemAcceptUrl(n.acceptedFrom.get, n))
              )

              val feedItemActions = Seq(acceptOrEditAction)
              n.copy(actions = feedItemActions)

            }.getOrElse{
              feedNewsitem
            }

          case _ =>
            feedNewsitem
        }
      }
    }

    Future.sequence(feedNewsitems.map(addFeedItemsActions))
  }

}
